src/neoxp/Node/Modules/ExpressRpcMethods.cs

```diff
@@ -403,6 +405,90 @@ public JObject GetNep11Properties(JArray @params)
             }
             return json;
         }
+        
+        private const byte Prefix_Contract = 8;
+        private const byte Prefix_NextAvailableId = 15;
+
+        private int LastUsedContractId(DataCache snapshot)
+        {
+            StorageKey key = new KeyBuilder(NativeContract.ContractManagement.Id, Prefix_NextAvailableId);
+            StorageItem item = snapshot.TryGet(key);
+            return (int)(BigInteger)item;
+        }
+
+        private void SetLastUsedContractId(DataCache snapshot, int newId)
+        {
+            StorageKey key = new KeyBuilder(NativeContract.ContractManagement.Id, Prefix_NextAvailableId);
+            StorageItem item = snapshot.GetAndChange(key);
+            item.Set(newId);
+        }
+        
+        private int GetNextAvailableId(DataCache snapshot)
+        {
+            StorageKey key = new KeyBuilder(NativeContract.ContractManagement.Id, Prefix_NextAvailableId);
+            StorageItem item = snapshot.GetAndChange(key);
+            int value = (int)(BigInteger)item;
+            item.Add(1);
+            return value;
+        }
+        
+        [RpcMethod]
+        public JObject? ExpressPersistContract(JObject @params)
+        {
+            var state = @params[0]["state"];
+            var storagePairs = (JArray) @params[0]["storage"];
+            var snapshot = neoSystem.GetSnapshot();
+
+            var nef = new NefFile
+            {
+                CheckSum = (uint)state["nef"]["checksum"].AsNumber(),
+                Compiler = state["nef"]["compiler"].AsString(),
+                Source = state["nef"]["source"].AsString(),
+                Tokens = new MethodToken[0], // TODO: support parsing MethodTokens or use fix of https://github.com/neo-project/neo/issues/2674
+                Script = Convert.FromBase64String(state["nef"]["script"].AsString())
+            };
+            
+            ContractManifest m = ContractManifest.FromJson(state["manifest"]);
+            
+            // Our local chain might already be using the contract id of the pulled contract, we need to check for this
+            // to avoid having contracts with duplicate id's. This is important because the contract id is part of the
+            // StorageContext used with Storage syscalls and else we'll potentially override storage keys or iterate
+            // over keys that shouldn't exist for one of the contracts.
+            int contractId = (int)state["id"].AsNumber(); 
```

Comment:
Shouldn't the offline implementation of contract download also be doing something like this? We need to consolidate the core logic for adding the downloaded contract/storage in one place so it's consistent between the two. This method should only be parsing the JSON and calling the shared function

Reply:
FYI, the `ExpressOracle` class provides static methods that are shared by on and offline node code. That might actually be the most logical place to put the common contract persistence logic, though we'd want to rename the class to something like `ExpressNodeCommon`

Reply:
want me to rename it as part of this PR?